apps/aecore/src/aec_conductor.erl

```diff
@@ -240,17 +242,18 @@ handle_call({add_synced_block, Block},_From, State) ->
     {reply, Reply, State1};
 handle_call(get_key_block_candidate,_From, #state{beneficiary = undefined} = State) ->
     {reply, {error, beneficiary_not_configured}, State};
-handle_call(get_key_block_candidate,_From, State) ->
+handle_call(get_key_block_candidate,_From,
+            #state{top_block = TopBlock, top_block_hash = TopBlockHash} = State) ->
     {Res, State1} =
         case State#state.pending_key_block of
             undefined ->
-                get_pending_key_block(aec_chain:top_block_hash(), State);
+                get_pending_key_block(TopBlock, State);
```

Comment:
This looks like a change in behavior? Using the local state rather than the current chain state?

Reply:
When a new block is successfully added, then the `TopBlock` is changed in `preempt_if_new_top`. I assume the `TopBlock` from the state and `aec_chain:top_block` and `top_block_hash` always refer to the `TopBlock`.

Reply:
The conductor is a program running in parallel with other stuff. I can imagine a race in which the top block has changed between the state update and the call to get the key block. Don't recall whether there is a reason exactly here, but in general we have had issues with concurrent updates in the past.

Reply:
My understanding is that even if there are several workers that are able to add new blocks (sync, gossip, HTTP API,...), all of these components do it using the conductor process, they just use `post_block` or `add_synced_block` functions from the conductor - these are calls. The conductor just needs to make sure each new added block is valid and connected to the genesis. Based on this, I thought it'd safe to just use the `TopBlock` from the state, not the one read from the db, as they are always the same. If there is something else that can update the db, I can just read the `TopBlock` from the db.